lrtable/src/lib/statetable.rs

```diff
@@ -306,10 +305,45 @@ where
         })
     }
 
+    fn decode(bits: u32) -> Action<StorageT> { // XXX howto find StorageT? encode in int as well?
+        //           Action   State/Prod
+        let mask = 0b00000000_11111111;
+        let action = bits >> 8;
+        let reference = bits & mask;
+
+        if action == 1 {
+            return Action::Shift(StIdx::from(reference))
+        }
+        else if action == 2 {
+            return Action::Reduce(PIdx(StorageT::from(reference).unwrap()))
+        }
+        else if action == 3 {
+            return Action::Accept
+        }
+        else {
+            return Action::Error
+        }
+    }
+
+    fn encode(action: Action<StorageT>) -> u32 {
+        let mut enc:u32 = 0;
+        match action {
+            Action::Shift(stidx) => { enc = (1 << 8) | (u32::from(stidx)); },
```

Comment:
This looks dodgy: surely one needs to do something like `u32::from(stidx) << 8`, otherwise the tag (`1<<8`) will be or'd with `stidx`?

Reply:
I'm not sure what you mean, `1<<8` is the first part of the `u32` encoding the actions, I then `or` this with the second part which is `stidx`.

Reply:
Ah, yes, I was wrong. Though it means you can only have an 8 bit `stidx`, so I think this code won't work with e.g. the java7 grammar.

Reply:
So shall we use a `u64` and use the first 2 bits for the `Action` and the remaining 62 bits for the states? Or use a `u64` and use the first 32 bits for the action and the remaining 32 for the states, to keep the values low. I guess since we say `StorageT` is maximum `u32` (and internally even `u16`) 32 bits for the states should be enough?